pkgs/dart_model/lib/src/json.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_changes.dart';
+
+/// JSON data.
+final class Json {
+  Map<String, Object?> _root;
+
+  Json.fromJson(Map<String, Object?> json) : _root = json;
+
+  JsonChanges computeChangesFrom(Json previous) {
+    final updates = <Update>[];
+    final removals = <Removal>[];
+    _compute(previous._root, _root, Path([]), updates, removals);
+    return JsonChanges(updates: updates, removals: removals);
+  }
+
+  static void _compute(
+      Map<String, Object?> previous,
+      Map<String, Object?> current,
+      Path path,
+      List<Update> updates,
+      List<Removal> removals) {
+    for (final key in previous.keys.followedBy(current.keys).toSet()) {
+      final keyIsInPrevious = previous.containsKey(key);
+      final keyIsInCurrent = current.containsKey(key);
+
+      if (keyIsInPrevious && !keyIsInCurrent) {
+        removals.add(Removal(path: path.followedByOne(key)));
```

Comment:
Do we plan to have `null` be a meaningful value in the schema for anything? If not, we could merge the Update/Removal classes and just have the value be null instead for removals. Or, even go so far as removing JsonChanges and representing the diff as Json with null entries for removals.